Transactor interface contract backend

anagha-ks opened the pull request:
<!-- Provide a general summary of your changes in the title above

Please read our contribution guidelines and sign the Contributor License
Agreement (CLA) before submitting the pull request. Also, check if there are no
other open pull requests targeting the same issue. -->

#### Description
Update perun-node to adopt the below added features in go-perun

- ChannelProposal modularization

- noapp nodata feature

- Transactor interface 

##### Category
<!-- Tell us what type of issue does your pull request target.
You can uncomment one of the following options: -->

Improvement

##### Relevant issue
Closes https://github.com/hyperledger-labs/perun-node/issues/105

#### Testing
All existing tests are passing
##### Steps to run the tests
<!-- Describe a set of steps to run the tests relevant to this change. -->
Nil

#### Checklist 
<!-- Please check if the pull request fulfils these requirements: -->

- [x] Name is added to the NOTICE file, if it is not present already.
- [x] Changes are rebased onto the target branch.

anagha-ks committed:
Update perun-node to use transactor interface

Update perun-node to use the transactor interface for sending the
keystore instead of using keystore directly at the contract backend.

Closes #105

Signed-off-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>

manoranjith commented:
Can you add a `Closes #105` phrase in the last commit message.

manoranjith commented on session/session.go:
```diff
@@ -255,7 +258,7 @@ func (s *session) OpenCh(
 	}
 
-	ch := newChannel(pch, openingBals.Currency, parts, s.timeoutCfg, challengeDurSecs)
+	ch := newChannel(pch, openingBals.Currency, []string{perun.OwnAlias, peer.Alias}, s.timeoutCfg, challengeDurSecs)
```
Can you use like before, declaring the `parts` and `partsAddress` variables next to each other, so that it is visually easier to ensure that the order or participants is same on both the arrays ?

anagha-ks replied:
Updated to old structure of declaring `parts`  and `partsAddress` variable in this [commit.](https://github.com/hyperledger-labs/perun-node/pull/110/commits/419c09f1a62eb87b03be5594f43da29dfd7ed31c)

manoranjith commented on blockchain/ethereum/chain.go:
```diff
@@ -52,6 +53,13 @@ func NewChainBackend(url string, chainConnTimeout, onChainTxTimeout time.Duratio
 	}
-	cb := pethchannel.NewContractBackend(ethereumBackend, ks, &acc)
+
+	// Create keystore wallet
```
```suggestion
```

anagha-ks replied:
Deleted in this [commit](https://github.com/hyperledger-labs/perun-node/pull/110/commits/419c09f1a62eb87b03be5594f43da29dfd7ed31c)

manoranjith commented on session/session.go:
```diff
@@ -235,13 +235,16 @@ func (s *session) OpenCh(
-		partAddrs,
-		pclient.WithApp(app.Def, app.Data),
-		pclient.WithRandomNonce())
+	proposal := &pclient.LedgerChannelProposal{ // Make proposal.
```
Can you use the `pclient.NewLedgerChannelProposal` function to construct the channel proposal, so that you are also able to pass options like `WithRandomNonce()`.

anagha-ks replied:
[Updated](https://github.com/hyperledger-labs/perun-node/pull/110/commits/419c09f1a62eb87b03be5594f43da29dfd7ed31c)

manoranjith replied:
Can you move that change from [add transactor interface](https://github.com/hyperledger-labs/perun-node/pull/110/commits/419c09f1a62eb87b03be5594f43da29dfd7ed31c) commit to this one, as it is a part of this change set ?

anagha-ks replied:
[Moved](https://github.com/hyperledger-labs/perun-node/pull/110/commits/03c61bdc7d403de9979c2b71a061e4521870b65a)

manoranjith commented on app/payment/session_test.go:
```diff
@@ -127,12 +127,13 @@ func init() {
-				Balances: [][]*big.Int{{ownAmountRaw, peerAmountRaw}},
-			},
-		},
+		Proposal: &pclient.LedgerChannelProposal{
```
Is it possible to use the `pclient.NewChannelProposal` function here ?

anagha-ks replied:
Since it is to fill the mock data structure in the notification, keeping as `&pclient.LedgerChannelProposal` itself.

manoranjith commented on perun.go:
```diff
@@ -253,7 +253,7 @@ type (
 		ProposalID string
 		Currency   string
-		Proposal   *pclient.ChannelProposal
+		Proposal   pclient.ChannelProposal
```
Since this interface now has a `Proposal` method that retrieves the underlying Proposal struct, is it possible to rename this, so that we don't use `Proposal.Proposal()` ?

anagha-ks replied:
Yes, I have changed `Proposal` to `ChProposal` which is addressed in this [commit](https://github.com/hyperledger-labs/perun-node/pull/110/commits/419c09f1a62eb87b03be5594f43da29dfd7ed31c)

manoranjith replied:
Can you move that change from [add transactor interface](https://github.com/hyperledger-labs/perun-node/pull/110/commits/419c09f1a62eb87b03be5594f43da29dfd7ed31c) commit to this one, as it is a part of this change set ?

anagha-ks replied:
[Updated](https://github.com/hyperledger-labs/perun-node/pull/110/commits/03c61bdc7d403de9979c2b71a061e4521870b65a)

manoranjith commented on app/payment/session.go:
```diff
@@ -69,7 +69,7 @@ func OpenPayCh(pctx context.Context,
 	challengeDurSecs uint64) (PayChInfo, error) {
 	paymentApp := perun.App{
-		Def:  ppayment.AppDef(),
+		Def:  ppayment.NewApp(),
```
Looking at the commit message, I think this change belongs to the [previous commit](https://github.com/hyperledger-labs/perun-node/pull/110/commits/88b3a7ff60ebb42d8f8f0ab10ada34ae97d0da99) that is related to change in `go-perun`.  Can you move the `payment.AppDef()` changes (here and all places below) to that commit.

So that one contains changes related only to the use of Transactor interface ?

anagha-ks replied:
Moved to this [commit](https://github.com/hyperledger-labs/perun-node/pull/110/commits/f955e9f33a5a09559f837b42cee4bc925859112e)

anagha-ks committed:
Update perun-node to noapp nodata feature

Update perun-node to adopt the noapp nodata changes made in go-perun

Signed-off-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>

anagha-ks committed:
Update perun-node to get ChannelProposal interface

This commit will update perun-node to get the channel proposal interface
feature introduced in go-perun.

Signed-off-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>
